sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FlattenRunner.java

```diff
@@ -0,0 +1,99 @@
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import org.apache.beam.fn.harness.data.BeamFnDataClient;
+import org.apache.beam.fn.harness.data.MultiplexingFnDataReceiver;
+import org.apache.beam.fn.harness.fn.ThrowingRunnable;
+import org.apache.beam.fn.harness.state.BeamFnStateClient;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Coder;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.runners.core.DoFnRunner;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.apache.beam.sdk.fn.data.FnDataReceiver;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.util.WindowedValue;
+
+/**
+ * A {@link DoFnRunner} for flatten transformations.
+ */
+public class FlattenRunner<InputT>{
+  /**
+   * A registrar which provides a factory to handle Java {@link DoFn}s.
+   */
+  @AutoService(PTransformRunnerFactory.Registrar.class)
+  public static class Registrar implements
+      PTransformRunnerFactory.Registrar {
+
+    @Override
+    public Map<String, PTransformRunnerFactory> getPTransformRunnerFactories() {
+      return ImmutableMap.of(
+          PTransformTranslation.FLATTEN_TRANSFORM_URN, new Factory());
+    }
+  }
+
+  /** A factory for {@link FlattenRunner}. */
+  static class Factory<InputT> implements
+      PTransformRunnerFactory<FlattenRunner<InputT>> {
+    @Override
+    public FlattenRunner<InputT> createRunnerForPTransform(
+        PipelineOptions pipelineOptions,
+        BeamFnDataClient beamFnDataClient,
+        BeamFnStateClient beamFnStateClient,
+        String pTransformId,
+        RunnerApi.PTransform pTransform,
+        Supplier<String> processBundleInstructionId,
+        Map<String, PCollection> pCollections,
+        Map<String, Coder> coders,
+        Map<String, RunnerApi.WindowingStrategy> windowingStrategies,
+        Multimap<String, FnDataReceiver<WindowedValue<?>>> pCollectionIdsToConsumers,
+        Consumer<ThrowingRunnable> addStartFunction,
+        Consumer<ThrowingRunnable> addFinishFunction)
+        throws IOException {
+
+      // Give each input a MultiplexingFnDataReceiver to all outputs of the flatten.
+      ImmutableSet.Builder<FnDataReceiver<WindowedValue<InputT>>> consumersBuilder =
+          new ImmutableSet.Builder<FnDataReceiver<WindowedValue<InputT>>>();
+      for (String output : pTransform.getOutputsMap().values()) {
+        consumersBuilder.addAll((Iterable) pCollectionIdsToConsumers.get(output));
+      }
+      Collection<FnDataReceiver<WindowedValue<InputT>>> consumers = consumersBuilder.build();
```

Comment:
drop this line and just call consumersBuilder.build() in the line below.

Reply:
Done.